Build failure on `v3.8.9`: `error: implicit declaration of function ‘locale_charset’`

There's also the error about returning int instead of const char, but I think that's just a result of it implicitly declaring the function with an int, which presumably is not supposed to happen.
```
[  0%] Building C object libarchive/CMakeFiles/archive.dir/archive_string.c.o
/home/chronos/user/libarchive/libarchive/archive_string.c: In function ‘default_iconv_charset’:
/home/chronos/user/libarchive/libarchive/archive_string.c:450:16: error: implicit declaration of function ‘locale_charset’ [-Wimplicit-function-declaration]
  450 |         return locale_charset();
      |                ^~~~~~~~~~~~~~
/home/chronos/user/libarchive/libarchive/archive_string.c:450:16: error: returning ‘int’ from a function with return type ‘const char *’ makes pointer from integer without a cast [-Wint-conversion]
  450 |         return locale_charset();
      |                ^~~~~~~~~~~~~~~~
make[2]: *** [libarchive/CMakeFiles/archive.dir/build.make:1045: libarchive/CMakeFiles/archive.dir/archive_string.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1281: libarchive/CMakeFiles/archive.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

What I think is the relevant section of the configuration log?
```
-- Performing Test HAVE_ICONV_libc_const
-- Performing Test HAVE_ICONV_libc_const - Failed
-- Performing Test HAVE_ICONV_libc_
-- Performing Test HAVE_ICONV_libc_ - Success
-- Looking for include file localcharset.h
-- Looking for include file localcharset.h - not found
-- Looking for locale_charset
-- Looking for locale_charset - found
```

Built as such: `mkdir builddir && cd builddir && cmake .. && make`.
Reproduced both on `v3.8.9` and latest master 19ff56da4f4790064346579a1a7f18a0230b0ac6.

--- a/libarchive/archive_string.c
+++ b/libarchive/archive_string.c
@@ -445,7 +445,7 @@ static const char *
 default_iconv_charset(const char *charset) {
 	if (charset != NULL && charset[0] != '\0')
 		return charset;
-#if HAVE_LOCALE_CHARSET && !defined(__APPLE__)
+#if HAVE_LOCALE_CHARSET && HAVE_LOCALCHARSET_H && !defined(__APPLE__)
 	/* locale_charset() is broken on Mac OS */
 	return locale_charset();
 #elif HAVE_NL_LANGINFO

